pkg/sql/opt/invertedidx/geo.go

```diff
@@ -61,6 +90,129 @@ func IsGeoIndexFunction(fn opt.ScalarExpr) bool {
+	factory *norm.Factory, fn opt.ScalarExpr, commuteArgs bool, inputProps *props.Relational,
+) (_ opt.ScalarExpr, inputGeoCol, indexGeoCol opt.ColumnID, ok bool) {
+	if !IsGeoIndexFunction(fn) {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on a function that cannot be index-accelerated",
+		))
+	}
+	function := fn.(*memo.FunctionExpr)
+
+	// Extract the the variable inputs to the geospatial function.
+	if function.Args.ChildCount() < 2 {
+		panic(errors.AssertionFailedf(
+			"all index-accelerated geospatial functions should have at least two arguments",
+		))
+	}
+
+	arg1, arg2 := function.Args.Child(0), function.Args.Child(1)
+	if commuteArgs {
+		arg1, arg2 = arg2, arg1
+	}
+
+	// The first argument should come from the input.
+	variable, ok := arg1.(*memo.VariableExpr)
+	if !ok {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on function containing non-variable inputs",
+		))
+	}
+	if !inputProps.OutputCols.Contains(variable.Col) {
+		return nil, 0, 0, false
+	}
+	inputGeoCol = variable.Col
+
+	// The second argument should be a variable corresponding to the index
+	// column. The caller must verify that this column is actually the key column
+	// of an inverted geospatial index.
+	variable, ok = arg2.(*memo.VariableExpr)
+	if !ok {
+		panic(errors.AssertionFailedf(
+			"TryGetInvertedJoinCondFromGeoFunc called on function containing non-variable inputs",
+		))
+	}
+	indexGeoCol = variable.Col
+
+	if commuteArgs {
+		// Copy the original arguments into a new list, and swap the first two
+		// arguments.
+		args := make(memo.ScalarListExpr, len(function.Args))
+		copy(args, function.Args)
+		args[0], args[1] = args[1], args[0]
+
+		// Get the geospatial relationship that is equivalent to this one with the
+		// arguments commuted, and construct a new function that represents that
+		// relationship.
+		rel := geoRelationshipMap[function.Name]
+		commutedRel := commuteGeoRelationshipMap[rel]
+		private := &function.FunctionPrivate
+		if rel != commutedRel {
+			// If the commuted relationship is not the same as the original
+			// relationship, we can't use the same function overload.
+			name := geoRelationshipReverseMap[commutedRel]
+			props, overload := findGeoFunction(name, args)
+			private = &memo.FunctionPrivate{
+				Name:       name,
+				Typ:        function.Typ,
+				Properties: props,
+				Overload:   overload,
+			}
+		}
+		fn = factory.ConstructFunction(args, private)
+	}
+
+	return fn, inputGeoCol, indexGeoCol, true
+}
+
+// findGeoFunction returns the function properties and overload of the
+// geospatial function with the given name and argument types matching the
+// given inputs. If no overload is found, findGeoFunction panics.
+func findGeoFunction(
+	name string, args memo.ScalarListExpr,
+) (*tree.FunctionProperties, *tree.Overload) {
+	props, overloads := builtins.GetBuiltinProperties(name)
+	for o := range overloads {
```

Comment:
Is it possible to use our standard overload code here? I'm not opposed to copy-paste, but if it's as easy as "just export a function and bam", we could consider using that. If it's difficult or impossible to do that, could you add a comment describing why?